pkg/controller/certificates/issuing/issuing_controller.go

```diff
@@ -238,14 +238,24 @@ func (c *controller) ProcessItem(ctx context.Context, key string) error {
 
 	cond := apiutil.GetCertificateRequestCondition(req, cmapi.CertificateRequestConditionReady)
 	if cond == nil {
+		// If the certificate request has been denied, set the last failure time to
+		// now, and set the Issuing status condition to False with reason. We only
+		// perform this check if the request also doesn't have a Ready condition,
+		// since some issuers may not honor a Denied condition, and will sign and
+		// set the Ready condition to True anyway. We would still want to complete
+		// issuance for requests where the issuer doesn't respect approval.
```

Comment:
Thanks for this thorough comment!!

I would suggest to focus on the "why" rather than "what" here:



```suggestion
		// Some issuers won't honor the Denied=True condition, and we don't
		// want to break these issuers. To avoid breaking these issuers, we skip
		// bubbling up the Denied=True condition from the request to the certificate
		// when the Ready condition unexpectedly appears on the request.
```

Reply:
We should document this behavior in cert-manager.io too

Reply:
Do we have a strategy for removing this exception in future? It'd be good to make it clear here whether it is intended to stay like this, and our expectations for new implementers of this API. (as well as documenting when implementations should be updated by)

Reply:
Hm come to think, why do we need this nested clause at all? Wouldn't simply having a check for `Denied` *always* be sufficient? If a signer does not actually honor the condition, I'm not sure if that's really an issue that this controller needs to handle?

Reply:
We also don't appear to be gating of the *Approved* condition being present either here, only failing if the Denied condition *is* present.

This could lead to non-deterministic behavior/race conditions depending on the order of controllers observing updates:

1) a CertificateRequest is created
2) the signer (which does not honor Approved/Denied conditions) signs the CSR and sets `status.certificate`
3) cert-manager's certificate controller issues the certificate into the Secret (as we aren't explicitly checking for Approved)
4) an approval controller denies the request

In these instances, the order of (4) and (2) will dictate what happens, which is confusing and unclear. If we are checking for Denied, I think it also makes sense to conversely check for Approved too.

I'd love to see this written down and discussed in the doc so implementers and users don't have to dig into this area of code to understand the nuance here

Reply:
I'll add this to the design document mentioned below.

The intention here is that an end user would want to have a denied certificate to be retried after a period of time, as if it was Failed. This seems like the most logical thing to do- either the policy is wrong, or the cert spec is wrong. A user would either change the Spec to be policy conformant, or policy would change and the cert will retry by itself.

It is worth noting that, if we _don't_ remove the Issuing condition, then a user wouldn't be able to manually renew the cert, as that command is simply setting the Issuing condition.

The reason for checking only if the Ready condition doesn't exist, is that a non-conformant external issuer will ignore the denied condition and sign the certificate. In this case, I would assume it to be reasonable that the user would want cert-manager to be permissive and make the Certificate to complete issuance.

We can continue the discussion on the doc.

Reply:
Related design document https://github.com/jetstack/cert-manager/pull/3797

Reply:
This thread seems to have reached a consensus, is that right?